### ISSUE TITLE
ints in pod spec somehow become scientific notation

### ISSUE THREAD

**Issue by jetpackdanger** [canonical/operator#303]
I have a pod spec that does this:
```
            'MM_FILESETTINGS_MAXFILESIZE': config['max_file_size'] * 1048576,
```
yielding the following Mattermost error:
```
Error: failed to load configuration: failed to load: failed to unmarshal config with env overrides: 1 error(s) decoding:                                                                                                                                                                                      
* cannot parse 'FileSettings.MaxFileSize' as int: strconv.ParseInt: parsing "5.24288e+06": invalid syntax
```

Mysteriously:
```
[agnew(charm-k8s-mattermost)] kubectl get -n mm deployment.apps/mattermost -o yaml | grep name:\ MM_FILESETTINGS_MAXFILESIZE -A 1
        - name: MM_FILESETTINGS_MAXFILESIZE
          value: "5.24288e+06"
[agnew(charm-k8s-mattermost)] _
```

5.24288e+06 is 5242880, which is 5 * 1048576, so it's obvious where the value came from.

But, curiously:
```
[agnew(charm-k8s-mattermost)] python3
Python 3.8.2 (default, Apr 27 2020, 15:53:34) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import json
>>> config = {'max_file_size': 5}
>>> json.dumps({'MM_FILESETTINGS_MAXFILESIZE': config['max_file_size'] * 1048576})
'{"MM_FILESETTINGS_MAXFILESIZE": 5242880}'
>>> _
```

I think what's happening is that Juju is reading 5242880 into a float (because JS and therefore JSON don't have anything but floats, although they JS floats can hold IIRC up to 53 bits of integer).

It'd be interesting to see if YAML pod specs behave the same way.

I'm not sure this is specifically a bug in the operator framework, but integers are common enough that I have a feeling this may become a problem folks run into.

**Comment by jetpackdanger** [comment]
Filed a substantially similar report against Juju, here: https://bugs.launchpad.net/juju/+bug/1881227

**Comment by jameinel** [comment]
I'm not positive, but I'm pretty sure the operator framework is passing it into Juju as your above JSON rendering, which means it is up to Juju how that gets interpreted.
I think there has been code in the past about "maybe_int" to handle the "treat it like an int if it looks like one".
There is also issue #293 which may mean we pass it as YAML, though IIRC the YAML parsing of pod spec inside of Juju is handled by K8s yaml interpreter (which IIRC rewrites the content into JSON first and then reads it)

**Comment by chipaca** [comment]
Fixed in Juju 2.8.1. Please reopen if you need this worked-around for older jujuses.
